File: src/lowlevel/client/icefsoperators/icefsRead.cpp
/*
 * @Author: Tan90degrees tangentninetydegrees@gmail.com
 * @Date: 2023-03-30 04:19:29
 * @LastEditors: Tan90degrees tangentninetydegrees@gmail.com
 * @LastEditTime: 2023-05-09 14:52:24
 * @FilePath: /icefs/src/lowlevel/client/icefsoperators/icefsRead.cpp
 * @Description:
 *
 * Copyright (C) 2023 Tan90degrees <tangentninetydegrees@gmail.com>.
 */
#include <stdio.h>

#include <string>

#include "icefsClient.hpp"
#include "icefsClientUtils.hpp"

void IcefsClient::DoIcefsRead(fuse_req_t fuseReq, fuse_ino_t inode, size_t size,
                              off_t offset, struct fuse_file_info *fi) {
  ICEFS_PR_FUNCTION;

  struct fuse_bufvec bufv = FUSE_BUFVEC_INIT(size);

  switch (this->clientConfig.linkType) {
    case ICEFS_LINK_USE_GRPC: {
      icefsgrpc::IcefsReadReq req;
      icefsgrpc::IcefsReadRes res;
      grpc::ClientContext ctx;

      req.set_size(size);
      req.set_offset(offset);
      req.set_fh(fi->fh);

      grpc::Status status = gRpcClient->DoIcefsRead(&ctx, req, &res);
      if (status.ok() && !res.status()) {
        bufv.buf[0].mem = const_cast<char *>(res.data().c_str());
        bufv.buf[0].size = res.size();
        int32_t ret = fuse_reply_data(fuseReq, &bufv, FUSE_BUF_SPLICE_MOVE);
        // int32_t ret = fuse_reply_buf(fuseReq, res.data().c_str(), res.size());
        if (ret != ICEFS_EOK) {
          fuse_reply_err(fuseReq, ret);
          std::cout << "ERR:" << __FUNCTION__ << "fuse_reply_buf" << ret
                    << std::endl;
        }
      } else {
        fuse_reply_err(fuseReq, res.status() ? res.status() : EIO);
        ICEFS_PR_ERR_STATUS;
      }
      break;
    }

    case ICEFS_LINK_USE_THRIFT: {
      icefsthrift::IcefsReadReq req;
      icefsthrift::IcefsReadRes res;

      req.__set_size(size);
      req.__set_offset(offset);
      req.__set_fh(fi->fh);

      icefsThriftConn *thriftConn = thriftClientPool->GetIcefsThriftConn();
      thriftConn->thriftClient->DoIcefsRead(res, req);
      thriftClientPool->PutIcefsThriftConn(thriftConn);
      if (!res.status) {
        bufv.buf[0].mem = const_cast<char *>(res.data.c_str());
        bufv.buf[0].size = res.size;
        int32_t ret = fuse_reply_data(fuseReq, &bufv, FUSE_BUF_SPLICE_MOVE);
        // int32_t ret = fuse_reply_buf(fuseReq, res.data.c_str(), res.size);
        if (ret != ICEFS_EOK) {
          fuse_reply_err(fuseReq, ret);
          std::cout << "ERR:" << __FUNCTION__ << "fuse_reply_buf" << ret
                    << std::endl;
        }
      } else {
        fuse_reply_err(fuseReq, res.status);
        ICEFS_PR_ERR_STATUS;
      }
      break;
    }

    default:
      fuse_reply_err(fuseReq, EIO);
      break;
  }
}

File: src/lowlevel/server/icefsoperators/icefsRead.go
/*
 * @Author: Tan90degrees tangentninetydegrees@gmail.com
 * @Date: 2023-03-11 07:18:32
 * @LastEditors: Tan90degrees tangentninetydegrees@gmail.com
 * @LastEditTime: 2023-04-18 08:15:28
 * @FilePath: /icefs/src/lowlevel/server/icefsoperators/icefsRead.go
 * @Description:
 *
 * Copyright (C) 2023 Tan90degrees <tangentninetydegrees@gmail.com>.
 */
package icefsoperators

import (
	"context"
	"icefs-server/icefserror"
	pb "icefs-server/icefsgrpc"
	"icefs-server/icefsthrift"
	"syscall"
)

func (s *IcefsServer) doIcefsRead(reqSize uint64, offset int64, fh uint64) (status int32, resSize uint64, data []byte) {
	var size int
	var err error

	if ((reqSize & (s.logicalBlockSize - 1)) | (uint64(offset) & (s.logicalBlockSize - 1))) == 0 {
		data = s.getAlignedMem(reqSize)
		size, err = syscall.Pread(int(fh), data, offset)
		status = icefserror.IcefsStdErrno(err)
		resSize = uint64(size)
	} else {
		var readOffset int64 = 0
		var paddingHead uint64 = 0
		var padding uint64 = 0
		if offset >= int64(s.logicalBlockSize) {
			paddingHead = uint64(offset & (int64(s.logicalBlockSize) - 1))
			readOffset = offset - int64(paddingHead)
		}
		padding = paddingHead + s.logicalBlockSize - ((reqSize + paddingHead) & (s.logicalBlockSize - 1))

		data = s.getAlignedMem(reqSize + padding)
		size, err = syscall.Pread(int(fh), data, readOffset)
		status = icefserror.IcefsStdErrno(err)
		size -= int(padding)
		if size >= 0 {
			resSize = uint64(size)
		} else {
			resSize = 0
		}

		data = data[paddingHead : paddingHead+resSize]
	}
	return
}

func (s *IcefsGRpcServer) DoIcefsRead(ctx context.Context, req *pb.IcefsReadReq) (*pb.IcefsReadRes, error) {
	var res pb.IcefsReadRes

	res.Status, res.Size, res.Data = s.server.doIcefsRead(req.Size, req.Offset, req.Fh)

	return &res, nil
}

func (s *IcefsThriftServer) DoIcefsRead(ctx context.Context, req *icefsthrift.IcefsReadReq) (*icefsthrift.IcefsReadRes, error) {
	var res icefsthrift.IcefsReadRes
	var size uint64

	res.Status, size, res.Data = s.server.doIcefsRead(uint64(req.Size), req.Offset, uint64(req.Fh))
	res.Size = icefsthrift.Ui64(size)

	return &res, nil
}
